src/phoenix/server/encryption.py

```diff
@@ -0,0 +1,136 @@
+"""Encryption utilities for sensitive data storage using Fernet.
+
+Security Model:
+---------------
+This module provides encryption-at-rest for sensitive configuration data and secrets
+stored in the Phoenix database. The encryption uses Fernet (symmetric encryption based
+on AES-128-CBC with HMAC authentication).
+
+Key Derivation:
+- The encryption key is derived from the Phoenix server's secret using PBKDF2-HMAC-SHA256
+- A fixed public salt is used (see _ENCRYPTION_KEY_DERIVATION_SALT below)
+- The salt doesn't need to be secret; its purpose is to prevent rainbow table attacks
+  against the password/secret, and to ensure different installations derive different keys
+- 100,000 iterations are used (OWASP recommended minimum as of 2023)
+
+Threat Model:
+- Protects against: Database dumps falling into unauthorized hands
+- Does NOT protect against: Attackers who have access to both the database AND the server secret
+- Key rotation: Changing the server secret will make existing encrypted data unreadable.
+  There is currently no key rotation mechanism - this is a known limitation.
+
+Usage:
+- All encryption/decryption should go through the EncryptionService class
+- Never store plaintext sensitive data in the database
+- The is_encrypted() heuristic is a best-effort validation, not cryptographic proof
+"""
+
+import base64
+
+from cryptography.fernet import Fernet, InvalidToken
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
+from starlette.datastructures import Secret
+
+# Salt for key derivation
+# This is intentionally public and hardcoded. Each Phoenix installation will derive
+# a different key based on their unique server secret. The salt prevents rainbow
+# table attacks and ensures key uniqueness across installations.
+_ENCRYPTION_KEY_DERIVATION_SALT = b"phoenix-database-encryption-a5d354120d50240ac80b74f56d09b363"
+
+# Fernet token structure constants
+_FERNET_VERSION_BYTE = 0x80  # Fernet tokens always start with version byte 0x80
+_FERNET_MIN_TOKEN_SIZE = (
+    57  # Minimum decoded size: version(1) + timestamp(8) + IV(16) + ciphertext(>=1) + HMAC(32)
+)
+
+
+class EncryptionService:
+    """Service for encrypting and decrypting sensitive data."""
+
+    def __init__(self, secret: Secret | None = None) -> None:
```

Comment:
the defaulting to none when the secret is not set might raise some flags. I wonder if it's worth having a global entropy here in the case the secret is not set.

> Require a high-entropy secret (e.g. at least 128 bits of randomness).
	•	Optionally store a per-installation random salt instead of a fully hardcoded one, so an attacker cannot reuse precomputation across different deployments.